File: ios/Classes/BetterPlayerEzDrmAssetsLoaderDelegate.m
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "BetterPlayerEzDrmAssetsLoaderDelegate.h"

@implementation BetterPlayerEzDrmAssetsLoaderDelegate

NSString *_assetId;

NSString * DEFAULT_LICENSE_SERVER_URL = @"https://fps.ezdrm.com/api/licenses/";

//MGR: accept drmHeaders from flutter
- (instancetype)init:(NSURL *)certificateURL withLicenseURL:(NSURL *)licenseURL withDrmHeaders:(NSDictionary*)drmHeaders {
    self = [super init];
    _certificateURL = certificateURL;
    _licenseURL = licenseURL;
    //MGR: store drmHeaders received from flutter DRM configuration
    _drmHeaders = drmHeaders;
    return self;
}

/*------------------------------------------
 **
 ** getContentKeyAndLeaseExpiryFromKeyServerModuleWithRequest
 **
 ** Takes the bundled SPC and sends it to the license server defined at licenseUrl or KEY_SERVER_URL (if licenseUrl is null).
 ** It returns CKC.
 ** ---------------------------------------*/
- (NSData *)getContentKeyAndLeaseExpiryFromKeyServerModuleWithRequest:(NSData*)requestBytes and:(NSString *)assetId and:(NSString *)customParams and:(NSError *)errorOut {
    NSData * decodedData;
    NSURLResponse * response;
    
    NSURL * finalLicenseURL;
    if (_licenseURL != [NSNull null]){
        finalLicenseURL = _licenseURL;
    } else {
        finalLicenseURL = [[NSURL alloc] initWithString: DEFAULT_LICENSE_SERVER_URL];
    }
    //NSURL * ksmURL = [[NSURL alloc] initWithString: [NSString stringWithFormat:@"%@%@%@",finalLicenseURL,assetId,customParams]];
    //MGR: appending data to URL causes 404 so skip this part; content ID is part of preAuthorization header
    NSURL * ksmURL = [[NSURL alloc] initWithString: [NSString stringWithFormat:@"%@",finalLicenseURL]];

    NSMutableURLRequest * request = [[NSMutableURLRequest alloc] initWithURL:ksmURL];
    [request setHTTPMethod:@"POST"];
    [request setValue:@"application/octet-stream" forHTTPHeaderField:@"Content-type"];
    //MGR: set Authorization and PreAuthorization headers in license request
    for(id key in _drmHeaders)
                [request setValue: [_drmHeaders objectForKey:key] forHTTPHeaderField: key];
    [request setHTTPBody:requestBytes];
    
    @try {
        decodedData = [NSURLConnection sendSynchronousRequest:request returningResponse:&response error:nil];
    }
    @catch (NSException* excp) {
        NSLog(@"SDK Error, SDK responded with Error: (error)");
    }

    //MGR: response is a json with base64 encoded CKC
    NSDictionary *json = [NSJSONSerialization JSONObjectWithData:decodedData options:0 error:nil];
    NSString *encodedCkc = [json objectForKey: @"CkcMessage"];
    NSData *ckc = [[NSData alloc] initWithBase64EncodedString:encodedCkc options:0];
    NSLog(@"Received CKC %d", ckc.length);

    return ckc;
}

/*------------------------------------------
 **
 ** getAppCertificate
 **
 ** returns the apps certificate for authenticating against your server
 ** the example here uses a local certificate
 ** but you may need to edit this function to point to your certificate
 ** ---------------------------------------*/
- (NSData *)getAppCertificate:(NSString *) String {
    NSData * certificate = nil;
    NSURLResponse * response;
    // certificate = [NSData dataWithContentsOfURL:_certificateURL];
    //MGR: get the certificate from server using DRM headers
    NSURL * certURL = [[NSURL alloc] initWithString: [NSString stringWithFormat:@"%@",_certificateURL]];
    NSMutableURLRequest * request = [[NSMutableURLRequest alloc] initWithURL:certURL];
    [request setHTTPMethod:@"GET"];
    //MGR: it is enough to pass Authorization header but be generous and pass all DRM headers :-)
    for(id key in _drmHeaders)
        [request setValue: [_drmHeaders objectForKey:key] forHTTPHeaderField: key];

    @try {
        certificate = [NSURLConnection sendSynchronousRequest:request returningResponse:&response error:nil];
    }
    @catch (NSException* excp) {
        NSLog(@"SDK Error, SDK responded with Error: (error)");
    }
    NSLog(@"Received Certificate %d", certificate.length);
    return certificate;
}

- (BOOL)resourceLoader:(AVAssetResourceLoader *)resourceLoader shouldWaitForLoadingOfRequestedResource:(AVAssetResourceLoadingRequest *)loadingRequest {
    NSURL *assetURI = loadingRequest.request.URL;
    NSString * str = assetURI.absoluteString;
    NSString * mySubstring = [str substringFromIndex:str.length - 36];
    _assetId = mySubstring;
    NSString * scheme = assetURI.scheme;
    NSData * requestBytes;
    NSData * certificate;
    if (!([scheme isEqualToString: @"skd"])){
        return NO;
    }
    @try {
        certificate = [self getAppCertificate:_assetId];
    }
    @catch (NSException* excp) {
        [loadingRequest finishLoadingWithError:[[NSError alloc] initWithDomain:NSURLErrorDomain code:NSURLErrorClientCertificateRejected userInfo:nil]];
    }
    @try {
        requestBytes = [loadingRequest streamingContentKeyRequestDataForApp:certificate contentIdentifier: [str dataUsingEncoding:NSUTF8StringEncoding] options:nil error:nil];
    }
    @catch (NSException* excp) {
        [loadingRequest finishLoadingWithError:nil];
        return YES;
    }
    
    NSString * passthruParams = [NSString stringWithFormat:@"?customdata=%@", _assetId];
    NSData * responseData;
    NSError * error;
    
    responseData = [self getContentKeyAndLeaseExpiryFromKeyServerModuleWithRequest:requestBytes and:_assetId and:passthruParams and:error];
    
    if (responseData != nil && responseData != NULL && ![responseData.class isKindOfClass:NSNull.class]){
        AVAssetResourceLoadingDataRequest * dataRequest = loadingRequest.dataRequest;
        [dataRequest respondWithData:responseData];
        [loadingRequest finishLoading];
    } else {
        [loadingRequest finishLoadingWithError:error];
    }
    
    return YES;
}

- (BOOL)resourceLoader:(AVAssetResourceLoader *)resourceLoader shouldWaitForRenewalOfRequestedResource:(AVAssetResourceRenewalRequest *)renewalRequest {
    return [self resourceLoader:resourceLoader shouldWaitForLoadingOfRequestedResource:renewalRequest];
}

@end

File: ios/better_player/Sources/better_player/BetterPlayerEzDrmAssetsLoaderDelegate.swift
import Foundation
import AVFoundation

public class BetterPlayerEzDrmAssetsLoaderDelegate: NSObject, AVAssetResourceLoaderDelegate {
    public let certificateURL: URL
    public let licenseURL: URL?

    private var assetId: String = ""
    private let defaultLicenseServerURL = URL(string: "https://fps.ezdrm.com/api/licenses/")!

    public init(_ certificateURL: URL, withLicenseURL licenseURL: URL?) {
        self.certificateURL = certificateURL
        self.licenseURL = licenseURL
        super.init()
    }

    private func getContentKeyAndLeaseExpiryFromKeyServerModule(request spc: Data, assetId: String, customParams: String) -> Data? {
        let finalLicenseURL = licenseURL ?? defaultLicenseServerURL
        guard let ksmURL = URL(string: "\(finalLicenseURL.absoluteString)\(assetId)\(customParams)") else { return nil }
        var request = URLRequest(url: ksmURL)
        request.httpMethod = "POST"
        request.setValue("application/octet-stream", forHTTPHeaderField: "Content-type")
        request.httpBody = spc

        let semaphore = DispatchSemaphore(value: 0)
        var resultData: Data?
        let task = URLSession.shared.dataTask(with: request) { data, _, _ in
            resultData = data
            semaphore.signal()
        }
        task.resume()
        _ = semaphore.wait(timeout: .now() + 30)
        return resultData
    }

    private func getAppCertificate() throws -> Data {
        return try Data(contentsOf: certificateURL)
    }

    public func resourceLoader(_ resourceLoader: AVAssetResourceLoader, shouldWaitForLoadingOfRequestedResource loadingRequest: AVAssetResourceLoadingRequest) -> Bool {
        guard let assetURI = loadingRequest.request.url else { return false }
        let urlString = assetURI.absoluteString
        let scheme = assetURI.scheme ?? ""
        guard scheme == "skd" else { return false }

        if urlString.count >= 36 {
            let startIndex = urlString.index(urlString.endIndex, offsetBy: -36)
            assetId = String(urlString[startIndex...])
        }

        let certificate: Data
        do {
            certificate = try getAppCertificate()
        } catch {
            loadingRequest.finishLoading(with: NSError(domain: NSURLErrorDomain, code: NSURLErrorClientCertificateRejected))
            return true
        }

        let requestBytes: Data
        do {
            guard let contentIdData = urlString.data(using: .utf8) else {
                loadingRequest.finishLoading(with: nil)
                return true
            }
            requestBytes = try loadingRequest.streamingContentKeyRequestData(forApp: certificate, contentIdentifier: contentIdData, options: nil)
        } catch {
            loadingRequest.finishLoading(with: nil)
            return true
        }

        let passthruParams = "?customdata=\(assetId)"
        let responseData = getContentKeyAndLeaseExpiryFromKeyServerModule(request: requestBytes, assetId: assetId, customParams: passthruParams)

        if let responseData = responseData, !responseData.isEmpty {
            loadingRequest.dataRequest?.respond(with: responseData)
            loadingRequest.finishLoading()
        } else {
            loadingRequest.finishLoading(with: NSError(domain: NSURLErrorDomain, code: NSURLErrorBadServerResponse))
        }
        return true
    }

    public func resourceLoader(_ resourceLoader: AVAssetResourceLoader, shouldWaitForRenewalOfRequestedResource renewalRequest: AVAssetResourceRenewalRequest) -> Bool {
        return self.resourceLoader(resourceLoader, shouldWaitForLoadingOfRequestedResource: renewalRequest)
    }
}
